### DISCUSSION TITLE
C++14

### DISCUSSION THREAD

**Discussion by maxhgerlach** [horovod/horovod#2791] [Q&A]
Have you considered moving Horovod to `-std=c++14`? It comes with some ease-of-use enhancements over C++11. On [Compiler Explorer](https://godbolt.org/z/jxMh8Mf67) it looks as if there is support with g++ 4.9 and later.

**Comment by tgaddair** [comment]
Hey @maxhgerlach, good suggestion. I can't think of a good reason not to upgrade to C++ 14. @romerojosh @leezu any thoughts?

**Comment by leezu** [comment]
CentOS7 comes with gcc 4.8.5 and still receives updates until June 30th, 2024. CentOS7 may still be common in some enterprise environments, but alternatives with newer toolchain exist (such as Amazon Linux 2). Requiring g++ 4.9 would induce extra steps for users on CentOS7 (such as `yum install devtoolset-9` to obtain gcc9).

@tgaddair, you can decide if that's an issue. Personally, I think it's acceptable to require some extra steps for CentOS7 users. Another option is to adopt gcc 7 / C++17, as the gcc 7 is commonly available on non-EOL systems newer than CentOS7 (and can be obtained on CentOS7 via the devtoolset).

**Reply by maxhgerlach** [reply]
Ubuntu 16.04 will be out of support on May 1st. Maybe that could be a good occasion to update to C++17? Ubuntu 18.04 comes with gcc 7 out of the box while it's readily available via the toolchain PPA for users who don't update from 16.04 in time.
